Event basics (#265)
open-telemetry/oteps: text/0265-event-vision.md

# Event Basics

## Motivation

The introduction of Events has been contentious, so we want to document and agree on a few basics.

### What are OpenTelemetry Events?

OpenTelemetry Events are a type of OpenTelemetry Log that requires an event name and follows a specific structure implied by that event name.

They are a core concept in OpenTelemetry Semantic Conventions.

### OTLP

Since OpenTelemetry Events are a type of OpenTelemetry Log, they share the same OTLP log data structure and pipeline.

### API

> OpenTelemetry should provide a (user-facing) Log API that includes the capability to emit OpenTelemetry Events.

Reviewer: Apart from mentioning `Log` way more than I think this covers the "Event" portion of the discussion (whether or not I fully agree on it being combined 🥲 ).

However, I would "suggest" that the section should mention (or reference a new Issue perhaps) that as part of the (user-facing) Log API there is discussions required about set of capabilities that it will have.

Or maybe just say

OpenTelemetry should provide a (user-facing) Log API that includes the capability to emit OpenTelemetry Events `to enforce the basic general semantic conventions.`

"enforce" might be a bit harsh, so some other wording about "validating" or maybe just to provide context that, "ensuring that they are followed" etc.

Author: > However, I would "suggest" that the section should mention (or reference a new Issue perhaps) that as part of the (user-facing) Log API there is discussions required about set of capabilities that it will have.

added this to the open questions section

# Event basics

## Motivation

The introduction of Events has been contentious, so we want to write down and agree on a few basics.

### What are events?

> Events are a type of log which has a required event name and a specific structure which is implied by that event name.

Reviewer 2: I think this has been the core of the contention. I believe this was a bad decision, to conflate logs and events. Even though events can be viewed as syntactically equivalent to logs, the purpose seems quite different. 

IMO, events should be used for semantically defined _events_, as well as custom user-defined business-events, and not viewed from a user perspective as being just "a type of log". This makes the distinction between logs and events way to fuzzy, and will lead to user confusion.

I know I have argued this in the past and lost the argument, but I do think it's a mistake to treat events as logs, when from a user perspective, they should have a different semantic purpose.

Reviewer: I agree, this wording is the problem

> Events are a type of log which has a required event name

The reason that `events` are being treated as a LogRecord was
- to avoid introducing a new transport / protobuf
- Span Events didn't (and now won't) support nested attributes

As John mentions, events are NOT "logs" (they are only transported via Logs)

Reviewer 3: Maybe we need to take a step back and see if folks have the same understanding of "logs".

1. https://opentelemetry.io/docs/concepts/signals/logs/ "A log is a timestamped text record, either structured (recommended) or unstructured, with optional metadata."
2. https://www.crowdstrike.com/cybersecurity-101/observability/three-pillars-of-observability/ "Every activity that happens in your applications and systems generates an [event log](https://www.crowdstrike.com/cybersecurity-101/observability/event-log/) with details that include things like a timestamp, event type, and machine or user ID. Whether written in plain text or binary, [structured or unstructured](https://www.crowdstrike.com/cybersecurity-101/observability/structured-logging/), log data contains text and metadata that provide a level of granularity that’s valuable for debugging and gaining insights into system events and errors."
3. https://www.techtarget.com/searchitoperations/tip/The-3-pillars-of-observability-Logs-metrics-and-traces "Logs are files that record events, warnings and errors as they occur within a software environment. Most logs include contextual information, such as the time an event occurred and which user or endpoint was associated with it."

I can start with my understanding:

* I think the opentelemetry.io description ("is a ... text record") is not what OpenTelemetry envisions, it doesn't have to be limited to "text" at all.
* I think the TechTarget description is not what OpenTelemetry envisions, logs are not limited to files.
* I think the CrowdStrike description is better aligned with what we want to have in OpenTelemetry.

Reviewer 4: my POV:
- logs (unstructured or semi structured blobs of text with no semantics) is a legacy, almost everything emits them, but we don't people to use them
- events (structured things with documented semantics) is the future, we want everyone to start using them

Existing logging facades usually do something in between - you can use logging APIs in a structured or unstructured way.

By saying that events are logs v2, we're trying to move everyone to a structured data. We're saying that OTel will (where necessary) provide the better-than-logs thing. We'll discourage using logs for new projects based on otel. But we'll stay compatible with unstructured logs.

If we try to separate logs and events:
- we'll keep unstructured logs as one of the things we support 
- we'll keep existing level of confusion among ourselves and users (do I use log or event for that thing)?
- we'll not have a reasonable story for structured-semantical-documented-things you create with traditional logging

It seems all the difference about logs and events is individual perception. Since we have different perceptions within this community, we'd fail to explain the difference to users.

Reviewer 2: If events are "logs v2", I think we should not transport them over the same channel as "logs v1". They should be separated, so as not to cause confusion between the two. How is one to know if a given log record is a "v1 log" or a v2 log (i.e. "event")? You can easily use "v1 logs" structure to create precisely identical payloads to a "v2 log", and hence there is no practical way to distinguish them.

I would absolutely support events being "logs v2", and I actually think this is an excellent idea, but as long as they are co-mingled in the transport with "logs v1" records, I think we will never be able to resolve how they differ.

Reviewer 4: logs v2 support all that logs v1 support from data structure standpoint + more - is there any controversy in it?

We have a discriminating property - presence of the `event.name` attribute, we can in theory create other discriminating properties if that's not enough for some reason. 

if you can use traditional logs to create properly structured and named event, then such logs are already logs v2  - what's wrong with this?

Reviewer 4: to be more precise: Events API is Logs API v2, and we happen to have data model that's already ~perfect~ compatible with both.

Reviewer 2: If you can already use a traditional log API to generate logs v2, then what value are we adding? Events are nothing but semantic conventions for logs, then, and I think we should drop the name "events" entirely. Unless our events are really something unique from existing structured logs, let's skip them altogether.

Reviewer 3: > Unless our events are really something unique from existing structured logs, let's skip them altogether.

+1

Author: events are both structured and named (and the expectation is that the name implies the structure)

Reviewer 4: > If you can already use a traditional log API to generate logs v2, then what value are we adding? 

traditional log API prioritizes string bodies and varies in ability to record context.

You'd have a subpar experience reporting structured bodies using SLF4J, Log4J, .NET ILogger, and many other loggers that support structured logging. You'll have no luck with loggers that don't support it such as JUL and many others.

With structured loggers, we'd need some subtle conventions for logging APIs to achieve it. E.g. potential approach with slf4j

```
logger.atInfo()
 .addKeyValue("event.name", "com.foo.my-event-id")
 .addKeyValue("otel.log.body", myEventBody) // `otel.log.body` property is used as log record body by bridge api
 .log("something important") 
```

Events API/Log v2 API is necessary to provide the API capable of defining all sorts of events consistently in OTel. If we don't have it, we'll need to solve structured logging for each external logging facade we need to support.

Some logging facade (like .NET ILogger)  may change to accommodate events better, but most of them probably won't. 

---------------------------

TL;DR: call it Event API or Logging API (v2), we need to have some API in OTel that allows any otel-api user to create structured named events (aka logs).

A different name for this API is helpful so we don't need to explain that it's not string-based log with questionable structure.

Reviewer 5: How is this conversation resolved? It seems like dissenting voices are just getting steamrolled here.

Author: I probably resolved it based on the comment above because our events really are something unique from existing structured logs: https://github.com/open-telemetry/oteps/pull/265#discussion_r1736401546

Unfortunately the discussions around events have tended to spin out into many different directions which makes it hard to know when they are resolved, or if they are only partially resolved then what portions still need to be resolved, which is why I've requested https://github.com/open-telemetry/oteps/pull/265#issuecomment-2317929226

> since some of the discussions are quite long, instead of unresolving it may help instead to start a new discussion with a summary of remaining discussion points (along with a pointer to the prior discussion for full context)

Reviewer 5: Gotcha, thanks for the link. I had missed that comment.

> The event name and its implied structure make events much more suitable for observability compared to traditional logs.

Reviewer 6: ```suggestion
The event name and its implied structure make events much more suitable for observability compared to unstructured logs.
```

I am not sure if "traditional logs" = "unstructured logs" ?

Author: I renamed "traditional logs" to "generic logs"

they may also be structured logs which lack an "event name"

> Because of this, the OpenTelemetry project wants to encourage the transition from traditional logs to events.

Reviewer 6: wants to encourage the transition from unstructured logs to structured logs, and events is one way to achieve this goal.

^ is this the intent?

Reviewer 7: Events are not just structured logs, they're structured logs with semantic rigor. The name field is required, providing a reliable invariant for classifying the type of thing that happened.  The fields of the body are (often) governed by semantic conventions, such that for a given event name, consumers can rely on the schema.

### OTLP

Since events are a type of log, they share the same OTLP data structure and OTLP pipeline.

### API

> OpenTelemetry should have an Event API. This will help to promote the distinction between events and traditional logs,

Reviewer 6: > OpenTelemetry should have an Event API

I am unsure. OTel decided to *not* introduce a end-user callable logging API in the past for a reason. Has that reason become invalid now? Or we believe Event API offers something unique that existing logging APIs do not have?

I am more aligned with the wording: "OpenTelemetry may offer an Event API for end users to call", if the existing logging solutions are not suited for generating structured events.

Reviewer 6: Thanks Jack for additional clarification. For the below pseudo-code
```
logger = GetLogger("my-scope");
logger.log(eventname="my-event", severity, body, attributes)
```

Is the idea that, for every event with the name "my-event", the body and attributes will have a fixed schema? And who is enforcing it? In the above, the user can totally produce yet another event with the same name "my-event", but this time with a diff schema for body/attributes. But the same is the case for Event API too:
```
event_logger = GetEventLogger("my-scope");
event_logger.logEvent(eventname="my-event", severity, body, attributes)
```

How does event api enforce that for every "my-event", body/attributes follow the same schema?

Reviewer 7: > Is the idea that, for every event with the name "my-event", the body and attributes will have a fixed schema? And who is enforcing it?

Yes. No enforcement besides the semantic conventions as of now. Its feasible that one day there could be a collector processor which rejects events which don't match the known schemas of an allow list.

> How does event api enforce that for every "my-event", body/attributes follow the same schema?

Folks have speculated that eventually, something like [weaver](https://github.com/open-telemetry/weaver) could be used to generate code which enforces more rigor on the event schema leveraging the languages type system. I.e. weaver could generate an API built on top of the Event API which accepts parameters for the specific fields of the event body.

Reviewer: > Is the idea that, for every event with the name "my-event", the body and attributes will have a fixed schema? 

Yes - https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/events.md, specifically
- The event name uniquely identifies event structure / type of the payload (body) and the set of attributes.
- "Unless the same event.name exists on two events, anything in two event bodies is not comparable to each other."

And just because (today) you can "simulate" the sending of an event by including an attribute called `event.name` on a Log Record does not mean (or directly imply) that you should "think" of events as "JUST" a structured Log. An `event` may represent a "thing" that occurred and while it's technically possible to "transport" these via the Logs structure some consumers / applications may want or need to treat these `events` in a different manner to that of a Log record.

Also from the above semantic convention of an event
> Semantically, an Event is a named occurrence at an instant in time. It signals that __"this thing happened at this time"__ and provides additional specifics about the occurrence. Examples of Events might include things like uncaught exceptions, button clicks, user logout, network connection severed, etc.

Reviewer 6: Even if we auto-generate code and instrumentations use them, users still can use the raw API to produce events with same-name-but-diff-schema.

Consider this example where an Event is defined:
```
[LoggerMessage(LogLevel.Critical, "A `{productType}` recall notice was published for `{brandName} {productDescription}` produced by `{companyName}` ({recallReasonDescription}).")]
    public static partial void FoodRecallNotice(
        this ILogger logger,
        string brandName,
        string productDescription,
        string productType,
        string recallReasonDescription,
        string companyName);
```

And it is used as shown below:
```
logger.FoodRecallNotice(
    brandName: "Contoso",
    productDescription: "Salads",
    productType: "Food & Beverages",
    recallReasonDescription: "due to a possible health risk from Listeria monocytogenes",
    companyName: "Contoso Fresh Vegetables, Inc.");
```

This is an example of OTel .NET's existing Log.
1. It has EventName - "FoodRecallNotice"
2. Compiler forces user to pass all the fields, thereby enforcing schema.
3. Still won't user from producing another event of same name with diff. schema!

I believe this is what code-gen tool can potentially generate for each language. For example, Otel Semantic Convention define an Event named `FoodRecallNotice`, it can also generate the helper code to achieve the same. And Instrumentations/End users can chose to use that auto-generated helper method.

Even if the helper is not auto-generated, one can write-it by hand, and ask users to call this to enforce schema.
What would be some of the reasons we ask an end-user to use the Event API via the helper, as opposed to the one I shown above?

Reviewer 4: @cijothomas I was thinking that from .NET perspective, `ILogger` will just provide both - traditional logging and events - it already does, it's just not possible to write structured bodies yet. Same with C++ or  Rust where we could make events first-class citizens and encourage using structured named things.

Author: > encourage using structured named things

exactly

and encourage the use of events over traditional logs.

### Interoperability with traditional logs

> It should be possible to send events from the Event API to a traditional logging library (e.g. Log4j).

Reviewer 2: I personally think this is a terrible idea, which will lead to a tremendous amount of confusion about what events are to be used for. If an event is just a structured log entry, then I think we should dump events altogether, as they add zero value over structured logging.

Reviewer 7: We can't have an event API, encourage its use instead of traditional logging APIs (slf4j, log4j, etc), and not have mechanisms for the data to be printed to traditional places like the console, files, etc. Nobody would use the event API.  

>  If an event is just a structured log entry

Events [are not just structured log entries](https://github.com/open-telemetry/oteps/pull/265#discussion_r1734979381).

Reviewer 2: As I commented above, if events are "logs v2", then they should _not_ be put in the same stream of data as "logs v1", otherwise, there is no difference, and events _are_ simply structured v1 log entries, perhaps with some sort of semantic meaning (that is completely non-obvious once the event has been co-mingled into the v1 log stream).

Reviewer 2: If that is actually the final state of things, then I strongly suggest we don't call them "events" and we just call them "semantically defined structured logs".

Reviewer 3: > If that is actually the final state of things, then I strongly suggest we don't call them "events" and we just call them "semantically defined structured logs".

+1

Reviewer 4: would this proposal imply renaming existing Event API to `SemanticallyDefinedStructuredLogging` (😒 ) ? Or `Logging API` (super-confusing with logs v1) ?

I think we need to draw a distinction in the name since events represent a subset of logs. Not all logs are events.

Reviewer 4: @reyang is your concern about ILogger, C++ logger, etc where we can have first-class support for events as a part of logging and it makes no sense to provide a separate API?

I assume we could have the distinction in the logging API via `Logger.log` and `Logger.event` or anything else (`[Event]` attribute instead of `[LogMessage]` attribute in .NET). 

 What's the perfect story you'd like to have?

Author: Having a name for "structured logs with semantic rigor" would help a lot with education / messaging / promoting their importance.

And having an API for these things would give OpenTelemetry first-class API coverage for all types of telemetry that can be described via semantic conventions.

Given the ambiguity of word "event", and given that it's a reasonably close fit, I personally think co-opting it into a specific definition for "OpenTelemetry Event™" seems vaguely reasonable.

Reviewer 3: > What's the perfect story you'd like to have?

https://github.com/open-telemetry/opentelemetry-specification/pull/3167#pullrequestreview-1283987908

I think it'll be great if OpenTelemetry can influence the logging API design across programming languages/runtimes. More details here https://github.com/open-telemetry/opentelemetry-specification/pull/3167#pullrequestreview-1283987908

To give some examples, I think it'll be awesome if we can say something like:

1. Rust has well established logging APIs from https://github.com/tokio-rs/tracing, OpenTelemetry worked closely with the Rust and Tokio communities, these APIs were enhanced to cover all the features.
2. .NET runtime has well established ILogger API, it already supports 95% of the features from what OpenTelemetry would need. .NET Foundation worked closely with OpenTelemetry community to close the 5% feature gap and now it covers 100% of the scenarios and features.
3. C++ doesn't have standard logging API, OpenTelemetry made progress to ship the logging API, and this work eventually influenced the C++ Standard Committee, so we worked together and put the API as part of the C++ standard libraries.

I think it'll be bad if we have something like:

1. SLF4J only covers 80% of what OpenTelemetry needs. OpenTelemetry ended up inventing a new set of competing APIs, now we have to do bridges from both sides and customers have to pick which one to use.

Reviewer 8: > I personally think this is a terrible idea, which will lead to a tremendous amount of confusion about what events are to be used for. If an event is just a structured log entry, then I think we should dump events altogether, as they add zero value over structured logging.

+

> As I commented above, if events are "logs v2", then they should not be put in the same stream of data as "logs v1". Once the data is in the log stream, there is effectively no difference between v1 and v2 (aside from defined semantic attributes)

I totally agree. If we are trying to promote a new thing, which we believe is better than traditional logs for observability, why not just have this concept in the SDK as done for metrics/spans? We just happen to use the same exporter and share the same data model but that's it. 

> We can't have an event API, encourage its use instead of traditional logging APIs (slf4j, log4j, etc), and not have mechanisms for the data to be printed to traditional places like the console, files, etc. Nobody would use the event API.

I may be missing something, but I don't see this as a strong argument. If I'm a end user, and want to use OTel events and I understand they are not logs, why would I expect it to be in places where normal logs are? 

We maybe be better off stop relating events with logs. Just because they are transmitted the same way via OTLP, they are not the same.

Reviewer 7: > I think it'll be great if OpenTelemetry can influence the logging API design across programming languages/runtimes.

@reyang the problem I see with this is that we don't just want to enhance existing logging APIs designs to have coverage for the OpenTelemetry event concept. We also want to actively discourage the use of traditional logs. An SLF4J logger which allows you to produce OpenTelemetry events still has 80%+ API surface area dedicated to producing traditional logs. Its too easy to do the wrong thing. The story would be better if SLF4J added dedicated support for OpenTelemetry events AND deprecated everything else, but that's not realistic.

Even if SLF4J were to have a perfect OpenTelemetry event story, we still would be stuck with the downside of having an inconsistent API story: "to use OpenTelemetry, take a dependency on the OpenTelemetry metrics and trace API, and the SLF4j logging / event API". Not necessarily a deal breaker, but it is a downside. Its cleaner to say "to Use OpenTelemetry, take a dependency on the OpenTelemetry API (for tracing, metrics, and events)". 

>  If we are trying to promote a new thing, which we believe is better than traditional logs for observability, why not just have this concept in the SDK as done for metrics/spans? We just happen to use the same exporter and share the same data model but that's it.

@joaopgrassi I think @jkwatson is saying that events should be a new proto type, so we'd have tracing, metrics, logs, and events. The reason we shouldn't do this is because the data model / proto message structure for events and logs would be essentially identical.

> If I'm a end user, and want to use OTel events and I understand they are not logs, why would I expect it to be in places where normal logs are? I use the API to produce events, I configure my OTLP exporter, I get the logs in my back-end and that's it.

We want library authors to use opentelemetry APIs for all observability. This means choosing the event API over existing log APIs. Library authors won't make this decision if they view opentelemetry APIs as overly restrictive about where the data ends up. From an opentelemetry standpoint, we expect data to be sent to a network location over OTLP, but for the large majority of users for which opentelemetry is still a new concept, they'll continue to expect to be able to turn on debug logging for a library and see the data on the console. 

I feel quite strongly about this: **library authors will not elect to replace logging APIs with the event API unless events can be _easily_ routed to typical log frameworks**.

Author: > customers have to pick which one to use

this OTEP says that "OpenTelemetry will recommend that application developers also use the Event API over emitting events via a generic logging library".

Reviewer 4: In the perfect world, I'd love us to go work with runtimes, languages and popular logging facades and implement 'better logging' there. 

We will probably succeed in some places and fail in others. Problems:

1. It'll take years or decades or have proper event support in most languages and popular libs
    
    E.g. in Java, you'd find at least 3 logging APIs that people use:
    - Java Util Logging (built-in into the language, does not support any structure)
    - SLF4J (supports structure prioritizing string messages)
    - Log4J API (supports structure prioritizing string messages)
    
    They all are extremely conservative and despite existing for two+ decades, they at major version 2. (log4j started doing betas for v3 recently that improves support for structured logs even more). 
    Despite releasing v2 major version recently (with structured logs support), SLF4J didn't break any public APIs and still prioritizes string messages. 

2. In languages where multiple logging libs are popular, OTel will need to pick one to depend on
    Assuming we don't provide better-than-logs API, OTel instrumentation libraries that will exist for years will need to pick one of the external logging libraries to emit events. 

3. We'll need to develop all sorts of weird conventions on how to emit events with every logging lib out there. It'll be hard and error prone

Reviewer 4: WRT proto/SDK:

If we could go back in time and could name `LogRecord` an `Event` instead (not possible), things could be the following:

- Event API and SDK encourage structured semantical events. Event proto represents them on OTLP
- 3rd party logging API usually creates string-body with optional attributes and uses Event API as a facade. (we still need a way to forward Event API to traditional logging systems and should be cautious of loops)

I.e. logs API is supported on top of events, logging systems/files/stdout can be used as exporting destinations, but things in the middle are the same.  What's the benefit of separating them (and asking users to understand subtle differences, configure two pipelines, etc)?

Reviewer 2: If we try to define a new logging API for Java, I can guarantee with near 100% certainty, that it will never be used by anyone outside the community. Java has multiple logging APIs already, and convincing people to switch would be a futile task. That doesn't even touch on the issue that having a first class logging API will require having a large configuration surface area, including formatters, appenders, etc etc.

Reviewer 4: That's why we're not trying to introduce a new logging API, we're using Event API which is already there and is experimental in Java:
 - it does not need formatters since it does not need string formatting
 - it's configuration is OTel configuration (do we want config to disable tracer by name? now we'll add config that configures `EventLogger` by name and verbosity)
 - OTel appenders are exporters and we only need a way to forward Event API to traditional logging appenders

Reviewer 2: I think this is an unrealistic expectation. If we treat this as a type of logging, people are going to ask for formatters and appenders, and advanced configuration, to match what they are used to from existing logging libraries.

Reviewer 8: > this is definitely an option, to frame Events as a new type (as opposed to a subset of Logs) which just happen to share a transport mechanism with Logs

Seems we are going in circles, and I wonder why we haven't discussed this option. Events are a new thing, why do we want to consider it as a log which carries ages of expectations and existing features on its back. I feel this could be much simpler if we just say hey, OTel has events, here's the API how to use it.

Reviewer 7: > If we treat this as a type of logging, people are going to ask for formatters and appenders, and advanced configuration, to match what they are used to from existing logging libraries.

Having an Event API and saying that the Events are a type of logging doesn't require formatters appenders. We can accommodate this by providing capabilities to route Event API calls back to existing log frameworks and their rich ecosystems of formatters / appenders. This could be done via _an alternative_ implementation of the Event API. I.e. the user installs in a Logback / Log4j / JUL implementation of the OpenTelemetry Event API instead of the OpenTelemetry SDK. OR, the OpenTelemetry SDK routes events back to Logback / Log4j / JUL in addition to exporting them to a network location via OTLP.

Reviewer 4: I think we have 3 discussions in this thread:

1. are Event API and traditional logging API the same? 
    - I think no - events are very opinionated logs (structure is forced) https://github.com/open-telemetry/oteps/pull/265#issuecomment-2322236851
2. do we need proper structured logging API in otel 
    - based on today's spec discussion - probably yes
    - would this API be any different from events API??
3. do we need separate SDK/pipeline for events and logs
    - ?

Reviewer 9: I'm going to write down the anology on (2) Do we need a proper structured logging API in otel that I gave in the spec meeting.

I think the answer to this question is *yes*. I also agree with @jkwatson that we don't want to force users to give up existing logging solutions they use today.  However - I disagree that we should use this as a justification for not providing an API for users.

The analogy (borrowed from someone else) is a Ski resort.  Ski resorts have ski rental shops.   Many folks who Ski bring their own skis, or may rent skis from other rental shops. The Ski resort still has its own rental shop because some folks will need to rent skis.   This is because the goal of the Ski resort is for folks to Ski.  Rentals may be a necessary component, but the Ski resort doesn't really care if folks rent somewhere else or bring their own Skis.

OpenTelemetry is about correlated, causal telemetry.  We need to interoperate with Logs (Skis, if you will) no matter how people bring them.  We should also provide our own API, given the demand from users and the need for us to be a complete solution.  However, we need *first class* support for existing logging APIs.

This is the reality of OpenTelemetry and *any existing, well established o11y API in any ecosystem*.

Reviewer 4: @jsuereth Using this analogy, would our Ski resort be interested in encouraging specific types of equipment we rent/allow to use? I think we need to do it. 

It's totally valid to use 3rd party skis created 30 years ago, but our lifts and slopes are optimized for different use-case. We won't sell or rent skis based on 30 years old designs in our shops.

To get back to logs, our potential logging API should prioritize structure and discourage string formatting. We won't be able to provide great experience for all 3rd party logs out there. We don't need to support string formatting features since it's bad for correlation.

If we want to have structured logs API, I believe it should be the same API as events. Overloads that take event name produce named-structured-logs (aka events), overloads that don't take name produce regular structured logs.

Reviewer 9: @lmolkova Yes - I'd agree with that as well.

Reviewer 10: > If we want to have structured logs API, I believe it should be the same API as events. Overloads that take event name produce named-structured-logs (aka events), overloads that don't take name produce regular structured logs.

I see this as a distinction without difference. An event emitted through the event API shouldn't cause an error, it should just emit an untyped event, aka, a structured log.

Reviewer 10: "All Events are structured logs, not all structured logs are Events" but the Event API can create either.

Reviewer 4: >An event emitted through the event API shouldn't cause an error, it should just emit an untyped event, aka, a structured log.

did you mean a **log** emitted through event API ? 

If so, I agree with this, we should have a general-purpose structured logging API capable of emitting named structured logs which might be called Events or something else https://github.com/open-telemetry/oteps/pull/265#issuecomment-2329633446

Reviewer 10: I meant an event in the generic sense, which could also be called a log, yes.

events = things that happen
Events = events with a defined semantic meaning
Untyped Events = structured events with no defined semantic meaning?

I think it's more important we figure out the taxonomy of how we want things to work in otel vs. how they comport with other systems. Obviously you can have structured logs in log4j. You can even have your own internal schema for those structured logs, making them logically equivalent to OTel Events. The differences are the things I noted in https://github.com/open-telemetry/oteps/pull/265#issuecomment-2327290693; namely, context.

Reviewer 11: Respectfully, but this whole debate – do need to make logs? And if so, do we need an API? And if so, what do we call the API? I don't think this is where the conversation really is. Over the past two years we've laid those questions to rest. If there are objectors, I get it, naming is hard. But I don't want us unravelling those decisions. The point of this doc is not to throw them out, it is to try and explain them properly. 

That said, here are some additional details that can hopefully explain why we have an Event API but no Event data structure.

### Why have an event API?

Otel has two goals when it comes to logging: 

* Consume all of the end users' existing logs.
* Produce our own highly opinionated logs, which will replace span events.

These goals are at the opposite ends of the spectrum. When consuming existing logs we want to be as flexible as possible, when creating our own logs we want to be as rigorous as possible. That's why we have both the `LogBridge` API and the `Event` API.

We can't avoid having our own instrumentation API. We have a very specific concept, and we need an API that allows us to express it correctly. Using the general purpose LogBridge API does not suffice, and we can't take a dependency on third party logging frameworks.

### Why aren't events a separate signal with their own data structure?

OpenTelemetry already has two data structures for events: LogRecords and SpanEvents. These two separate data structures exist because there are two kinds of backends: tracing backends and logging backends. 

Unified backends are relatively new.  We expect these new unified backends to store all events as LogRecords, and to deprecate the SpanEvent data structure. 

The point is: data structures are a configuration issue. Some end users will want the events to be recorded as LogRecords. Other end users will want events to be recorded as SpanEvents. Others will want both. It completely depends on their setup.

### Should we call these logs or events?

Otel events are both structured logs and the successor to SpanEvents. So what should we call them?

We choose the name Events as our term for Otel structured logging because it avoids the most confusion. There are many kinds of "structured logging" already involved with both the LogBridge API and the Collector logging pipeline. The LogRecord data structure itself can be thought of as "Otel structured logging."  

On the other hand, the term "event" is less confusing. Not because there aren't many theoretical definitions for the meaning of event, but because there aren't many practical "event" systems that users are trying to integrate with. By referring to our specific structure as an event helps to avoid confusion with the other forms of structured logging that are present in every user's logging pipeline.

Reviewer 11: trying to summarize here and bring these threads together into a decision: https://github.com/open-telemetry/oteps/pull/265#issuecomment-2334806152

Author: Resolving this based on latest updates. If you'd like to continue discussing, please start a new thread with a summary of any remaining points you'd like to discuss. Thanks.

This allows users to integrate events from the Event API into an existing (non-OpenTelemetry) log stream.

> Note: If a user chooses to send events from the Event API to a traditional logging library, and they have

Reviewer: Yeah, I'd be happy making this stronger or even "ban" (or more strongly word) this scenario...

While I can see (if I squint a LOT) why someone would want to do this, but adding wording to say and any Event SDK (or processor) that does this MUST NOT populate the `event.name` attribute "to avoid" the duplication as called out here. Which yes would lead to `events` JUST becoming (and treated as) standard unstructured log records -- which may be fine for some backends / use cases.

Author: I don't think we want to drop the `event.name` since it's such a critical part of event definition

also chosen to send the logs from their traditional logging library to the OpenTelemetry Logging SDK, then they should
avoid sending events from the Event API directly to the OpenTelemetry Logging SDK since that would lead to duplicate
capture of events that were sent from the Event API.

> It should be possible to bypass the Event API entirely and emit Events directly via an existing language-specific logging libraries.

Reviewer 6: Thanks for re-iterating this part.

Reviewer: > It should be possible to bypass the Event API entirely and emit Events directly via an existing language-specific logging libraries.

I'd prefer to add some wording around this to highlight that this is a stop-gap and NOT the overall intention.
ie. that "today" it's technically possible to simulate an event via a log record, but instrumentations SHOULD not use this technique IF an Event API is available for the language.

Author: @MSNev I believe this is addressed below:

> OpenTelemetry will recommend that
[instrumentation libraries](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/glossary.md#instrumentation-library)
use the Event API over emitting events via a generic logging library in order to give users a simple and consistent
onboarding story that doesn't involve mixing the two approaches.
>
> OpenTelemetry will recommend that application developers also use the Event API over emitting events via a generic
logging library since it avoid accidentally emitting logs which lack an event name or are unstructured.

This helps reinforce the idea that events are just a specific type of log.

Note: Because of this, generic event processors should be implemented as Log SDK processors.

OpenTelemetry will recommend that
[instrumentation libraries](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/glossary.md#instrumentation-library)
use the Event API over emitting events via a traditional logging library in order to give users a simple and consistent
onboarding story that doesn't involve mixing the two approaches.

> OpenTelemetry will recommend that application developers also use the Event API over emitting events via a traditional

Author: a critical part of events is the "event name" (not just the structure).

I've renamed "traditional logs" to "generic logs" in the text to maybe(?) help.

logging library since it avoid accidentally emitting (non-event) traditional logs.

Also, recommending the Event API over emitting events via a traditional logging library makes for a clearer overall
OpenTelemetry API story - with first class user facing APIs for traces, metrics, and events,
all suitable for using directly in native instrumentation.

### SDK

The Event SDK needs to support two destinations for events:

* OpenTelemetry Logging SDK
* Language-specific logging libraries

## Trade-offs and mitigations

TODO

## Prior art and alternatives

TODO

## Open questions

> TODO

Reviewer 12: I am curious about the direction/intent for opting in/out of specific events. My organization creates user-defined business-events. Today, I use a logging api with the convention of setting `event.name`. This can get clunky when a logger is shared between logs and events. Maybe I want logs at a severity of error but events at a severity of info. Do I create 2nd logger or something else.

This concern might be answered somewhat by https://github.com/open-telemetry/oteps/commit/e9fce2245c1482225684a7cb33579460b6ecf5e0. If we have a dedicated event api then this problem should be solved.

But then I am curious about how on subscribes to specific events (not specific to user-defined events). Do we get everything, register instrumentation scopes (like is done for other signals), something else?. T What relationship does `severity` play here? Is it purely metadata on the record or does it impact what is exported?

Reviewer 11: Having a seperate Event API – which exists today in [the spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/event-api.md) – is intended to assist with this. It is then possible to write [event processors](https://github.com/open-telemetry/opentelemetry-specification/pull/4196) which select, modify, and filter events. Otel events include the usual concepts found in logs, such as severity, so it is possible to write processors which utilize those features.

In general, having a pipeline for Otel events that is separate from your existing application logging pipeline is helpful in many situations. This is our justification for providing a complete logging framework for our Otel-specific logging structure, AKA OTel events. Users can choose when and how they integrate Otel events with their existing logs, or optionally leave them completely separate.

Author: > I am curious about the direction/intent for opting in/out of specific events

This is definitely important. In addition to @tedsuo's remarks, I think it's also addressed by this (under Future Possibilities):

> The Event API will probably need an `IsEnabled` function based on severity level, scope name, and event name.

## Future possibilities

> The Event API will probably need an `IsEnabled` function based on severity level, scope name, and event name.

Reviewer: As an OTEP I'd rephrase this as 

"based on the event.name and other possible attributes / properties of the event" (rather than calling out the "LogRecord" specific properties)

Author: this sentence doesn't mention "LogRecord", so it seems generic to me already

# Event basics

## Motivation

> The introduction of Events has been contentious, so we want to write down and agree on a few basics.

Reviewer 8: ```suggestion
The introduction of Events has been contentious, so we want to write down and agree on a few basics.
```

Nit: I would word it as "has created confusion" or "is a source of constant confusion".. contentious has a bad connotation, like there was a heated argument/conflicts. But I'm not a native speaker so maybe it's all fine and just different cultural views :)

Author: > like there was a heated argument/conflicts

there have been 😄

### What are events?

Events are a type of log which has a required event name and a specific structure which is implied by that event name.

> The event name and its implied structure make events much more suitable for observability compared to generic logs.

Reviewer 8: It would be great if we had _something_ to backup this claim that events >>> logs for observability.

Reviewer 13: I think the structure referenced on the line above is the reason. Logs (unstructured) are more difficult to parse, index, and aggregate than events (structured).

Author: And the required "event name" makes it easy for backends/users to differentiate between different events.

Because of this, the OpenTelemetry project wants to encourage the transition from generic logs to events.

### OTLP

Since events are a type of log, they share the same OTLP data structure and OTLP pipeline.

### API

> OpenTelemetry should have an Event API. This will help to promote the distinction between events and generic logs,

Reviewer 8: Without trying to be pedantic, we do have an event api https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/event-api.md. I guess here you want to express that we need an event api that produces events according to the semconv? https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/events.md

Reviewer 13: @joaopgrassi i'm not sure what the distinction is that you're making. the current event API is used to create semconv-compliant events.

Reviewer 8: it's just because the text says: "OpenTelemetry should have an Event API" so "should" there is confusing

Author: > we do have an event api https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/event-api.md

yeah, but there isn't consensus (yet) that we should keep it

Reviewer 13: There is already https://github.com/open-telemetry/oteps/blob/main/text/0202-events-and-logs-api.md

IMO the onus is on the nays to submit an OTEP to remove the event API rather than on the yeas to affirm it. That said, I think it's fine to keep this line. I'd just change it to "should keep its" instead of "should have an"

and encourage the use of events over generic logs.

> ### Interoperability with generic logging libraries

Author: TODO how does event body interop with generic logging libraries? (including anyvalue bodies)

possibility: map event body to jsonified string "generic log" body

Reviewer 11: I suggest that in the generic case we follow the same rules we have for mapping event attributes and the event body to a span event.

Custom mappings for event features will be a common request. These requests are easily handled by event processors and the log bridge, as it is simple to write tiny plugins that do these conversions.

Reviewer 11: (btw, since this OTEP is meant to be a framing document, I suggest that we avoid defining the specifics for how Events and SpanEvents map to each other.)

It should be possible to send events from the Event API to a generic logging library (e.g. Log4j).
This allows users to integrate events from the Event API into an existing (non-OpenTelemetry) log stream.

Note: If a user chooses to send events from the Event API to a generic logging library, and they have

> also chosen to send the logs from their generic logging library to the OpenTelemetry Logging SDK, then they should

Reviewer 8: > send the logs from their generic logging library to the OpenTelemetry Logging SDK

This is I guess via the Log bridge?

Author: Not necessarily. The Event SDK could also talk directly to the Logging SDK.

> avoid sending events from the Event API directly to the OpenTelemetry Logging SDK since that would lead to duplicate

Reviewer 8: How we envision this working? Will events also leverage log bridges or will events be produced by the SDK (e.g., like metrics) and just happen to use the same otlp log exporter that log bridges use? This is a bit confusing to be honest.

Author: I'm trying not to overspecify in the OTEP. The key part of the proposal is that Events should go to the Logging SDK (where they will go through Logging Processors -> Logging Exporters).

capture of events that were sent from the Event API.

> It should be possible to bypass the Event API entirely and emit Events directly via an existing language-specific logging libraries.

Reviewer 8: It should be possible, but _do we want to encourage it?_ If not, why we mention it here, as if it's a "bonus" that this is possible? 🤔. At least, that's how I interpret this sentence.

Author: > It should be possible but _do we want to encourage it?_

see recommendation below

> If not, why we mention it here, as if it's a "bonus" that this is possible?

to be clear that in this proposal events are a subset of general logs

This helps reinforce the idea that events are just a specific type of log.

Note: Because of this, generic event processors should be implemented as Log SDK processors.

OpenTelemetry will recommend that
[instrumentation libraries](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/glossary.md#instrumentation-library)
use the Event API over emitting events via a generic logging library in order to give users a simple and consistent
onboarding story that doesn't involve mixing the two approaches.

OpenTelemetry will recommend that application developers also use the Event API over emitting events via a generic
logging library since it avoid accidentally emitting logs which lack an event name or are unstructured.

Also, recommending the Event API over emitting events via a generic logging library makes for a clearer overall
OpenTelemetry API story - with first class user facing APIs for traces, metrics, and events,
all suitable for using directly in native instrumentation.

### SDK

The Event SDK needs to support two destinations for events:

> * OpenTelemetry Logging SDK

Reviewer 8: Isn't this OTLP Logging exporter as "destination for events"?

Author: No, the destination (in this proposal) is the Logging SDK, so that they will still go through standard Logging Processors -> Logging Exporters

> * Language-specific logging libraries

Reviewer 8: why we need to support sending events to logging libraries? Why not only through the SDK, like span and metrics?

## Trade-offs and mitigations

TODO

## Prior art and alternatives

TODO

## Open questions

TODO

> ## Future possibilities

Reviewer 9: We should call out how events interact with Metric data model / metric events within the SDK.

# Event Basics

## Motivation

The introduction of Events has been contentious, so we want to document and agree on a few basics.

### What are OpenTelemetry Events?

OpenTelemetry Events are a type of OpenTelemetry Log that requires an event name and follows a specific structure implied by that event name.

They are a core concept in OpenTelemetry Semantic Conventions.

### OTLP

Since OpenTelemetry Events are a type of OpenTelemetry Log, they share the same OTLP log data structure and pipeline.

### API

OpenTelemetry should provide a (user-facing) Log API that includes the capability to emit OpenTelemetry Events.

### Interoperability with Generic Logging Libraries

> It should be possible to send OpenTelemetry Logs from the OpenTelemetry Log API to a generic logging library (e.g., Log4j).

Reviewer 14: From a customer point of view, I think it may be confusing, for example, when to use OTel Logging API mapped to Log4j vs when to use Log4j directly. It also seems contrary to what OTel recommended before, which is "continue to use your language logging library". Should we be explicit "continue to use your existing logging APIs except when you need more structure, etc."?

Reviewer 11: I would like us to back away from passing judgements on these APIs. We should definitely explain the use cases: appending logs from application loggers, and writing OTel instrumentation. And it is worth mentioning that events can be created using any method on our logger, so application developers can choose to integrate the OTel conventions into their application logging via their log appender.

But I don't think there is a need to recommend that developers should stay away from the API. That feels like a very confusing recommendation.

Reviewer 11: Specifically regarding the case where users want to loop Otel logs through their own logger and back to OTel – I'm not really sold on this being a common use case. Which means that we need to support it, but we don't need to make a big deal about it or make it front and center.

The common use case presented for wanting to loop the logs back through the application logger is to export them  to other log appenders. But it's easy to add more exporters to OTel, so this use case doesn't seem like much of a barrier in the long term.

As far as other kinds of log processing are concerned, mutating Otel events is actively bad and unhelpful. It feels unlikely that many developers will want to do this. Again, we can support it, but users with this problem will self identify and find the docs for it, we don't need to make this use case a top level concept.

This allows users to integrate OpenTelemetry Logs into an existing (non-OpenTelemetry) log stream.

> It should also be possible to bypass the OpenTelemetry Log API entirely and emit OpenTelemetry Logs (including Events)

Reviewer 6: ```suggestion
It should also be possible to bypass the OpenTelemetry Log API entirely and emit OpenTelemetry Events
```

OpenTelemetry Logs can be already emitted via bridge.

Author: I think I prefer the current language because it reinforces that Events are just a type of Log

directly via existing language-specific logging libraries, if that library has the capability to do so.

OpenTelemetry will recommend that
[instrumentation libraries](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/glossary.md#instrumentation-library)
use the OpenTelemetry Log API to emit OpenTelemetry Events rather than using other logging libraries to emit OpenTelemetry Events. This recommendation aims to provide users with a simple and consistent
onboarding experience that avoids mixing approaches.

> OpenTelemetry will also recommend that application developers use the OpenTelemetry Log API to emit OpenTelemetry Events instead of using another

Reviewer 6: wondering why is this recommendation required? If the 'other api' has support for emitting structured logs with event name, should Otel have reason to recommend one over the other? I think this is best left to individual language clients to make recommendations.

Not a blocker as I believe language clients can always make their recommendations, and this OTEP already has an explicit call out about by passing OTel Log API and leveraging language existing API.

Reviewer: The primary reason I can think of is because it would allow the adopter (application), vender (distro) to configure their usage of events in whatever manner they would like without needing to create duplicate code to intercept any Log API or Log Processor or Logging framework to be able to redirect the events they care about to do whatever they want with them...

Reviewer 6: Irrespective of whether OTel Log API or 3rd party API is used, routing can be done in exact same way right? Both produce a LogRecord with EventName in it, with no way to distinguish between who/how produced it.

Author: > If the 'other api' has support for emitting structured logs with event name, should Otel have reason to recommend one over the other? I think this is best left to individual language clients to make recommendations.

are you thinking of the case (e.g. Rust and .Net), where the "other API" will be explicitly adopted as _the_ OpenTelemetry Log API? if so, that will be the one we recommend (since it will be _the_ OpenTelemetry Log API)

Reviewer 6: Only OTel .NET has make such adoption. OTel Rust **recommends** a particular existing logging API, nothing more. (so far). 

In case of OTel Rust/others, if the existing API can very well produce OTel Events (LogRecord with name and schema), any reason why we should recommend OTel owns API over other?

Logging libraries get improved all the time, so even the ones that are not good today may improve over time and might very well gain the ability to produce OTel Events. eg: Golang's native logging became `slog` and added structured logging recently. .NET's ILogger gained ability to automatically add EventName recently. 

Hence the qn- why OTel as a whole has to recommend our own API, vs leaving that to individual language clients?

If Rust's logging library evolves like that, and we can totally use that in instrumentations we maintain. I actually prefer to recommend people use Rust's existing logging library then instead of OTel's. The native logging libraries have other advantages that OTel probably don't want to chase and match.

logging library, as this helps prevent accidentally emitting logs that lack an event name or are unstructured.

> Recommending the OpenTelemetry Log API for emitting OpenTelemetry Events, rather than using a generic logging library, contributes to a clearer overall

Reviewer 3: What does "generic logging library" mean here? For example, OpenTelemetry Rust recommends users to use tokio: https://github.com/open-telemetry/opentelemetry-rust?tab=readme-ov-file#project-status. Does this mean OpenTelemetry Rust and Tokio need to change the direction and start to compete? (I don't think that's what we want)

Author: my understanding of the Rust case is that OpenTelemetry Trace API == tokio, and so similarly OpenTelemetry Log API == tokio

Reviewer 11: Yeah, from the perspective of the spec I don't think we should reference issues that only appear in the languages that plan to deviate from the spec due to issues specific to their ecosystem.

> OpenTelemetry API story. This ensures a unified approach with first-class user-facing APIs for traces, metrics, and events,

Reviewer 3: I can understand where this is coming from, I feel this is still leaving a long-term issue.

```suggestion
OpenTelemetry API story. This ensures a unified approach with first-class user-facing APIs for traces, metrics, and logs,
```

Author: "events" here is intentional 😅 

the suggestion is:
* if you are emitting OpenTelemetry Events - we recommend using the OpenTelemetry Log API (but you can also use a generic logging library if you want)
* if you are emitting other types of OpenTelemetry Logs - you can use either the OpenTelemetry Log API or a generic logging library (and we don't have any specific recommendation)

Reviewer 3: I know and that's why I said "I can understand ...". However, I think this is going to leave the confusion - similar to what I explained in https://github.com/open-telemetry/oteps/pull/265#discussion_r1757233167

Reviewer 3: > * if you are emitting other types of OpenTelemetry Logs - you can use either the OpenTelemetry Log API or a generic logging library (and we don't have any specific recommendation)

I think the confusion here is "what is a generic logging library". For example, Tokio might be considered the most popular "generic logging library" for Rust, throughout the collaboration between Tokio and OpenTelemetry Rust maintainers, it has made several improvements (e.g. adding Event Name support), which makes it fully ready for "events". My point is that things are evolving, whatever we call as "generic logging library" today can evolve into "100% OTel Logging API spec compatible library", and if we don't have clear understanding/explanation, users will be very confused.

Author: makes sense. I've updated everywhere to change "generic logging library" to "other logging library"

all suitable for direct use in native instrumentation.

### Relationship to Span Events

Events are intended to replace Span Events in the long term.

> Span Events will be deprecated to signal that users should prefer Events.

Reviewer 6: is the reason for this to avoid 2 things to achieve the same goal? SpanEvents get a nice property automatically by sampling in/out together with the span. I guess OTel Events will also get similar capability, at which point we SpanEvent has no particular advantage?

Author: > is the reason for this to avoid 2 things to achieve the same goal?

yes

> SpanEvents get a nice property automatically by sampling in/out together with the span. I guess OTel Events will also get similar capability

good point, I added sampling to the open questions

Interoperability between Events and Span Events will be defined in the short term.

### SDK

This refers to the existing OpenTelemetry Log SDK.

## Trade-offs and mitigations

> TODO

Reviewer 15: I think that the trade-off is overloading of the "events" term which is already used in OpenTelemetry (Span Events as mentioned before).

The issue is that people may start confuse span events with log events.

What is the mitigation for avoiding the confusion among the users?

Author: isn't that addressed by this section?

> ### Relationship to Span Events
>
> Events are intended to replace Span Events in the long term. Span Events will be deprecated to signal that users should prefer Events.
>
> Interoperability between Events and Span Events will be defined in the short term.

Reviewer 15: So maybe it will be better to move this section here.

Author: removed the empty sections

Reviewer: > I think that on of the main trade-offs is overloading of the "events" term which is already used in OpenTelemetry for Span Events (as mentioned before).
> The issue is that people may start confuse span events with log events.

The long term point and the Sem Conv tooling is that there are and will be the same, so it becomes a transport (HOW) the event will be transmitted (eventually), because an event is an event.

## Prior art and alternatives

TODO

## Open questions

* How to support routing logs from the Log API to a language-specific logging library

>   while simultaneously routing logs from the language-specific logging library to an OpenTelemetry Logging Exporter?

Reviewer 15: Someone writes a Log API implementation which is actually a log appender/sink/handler for given language-specific logging library.

Author: I'd prefer to be overgenerous in adding open questions / future possibilities (and not worry about them until we agree on the basic vision).

> * How do log bodies interoperate with generic logging libraries?

Reviewer 15: Depends on the logging libraries 😉 
Some support only string, some structured bodies.
I am not sure if this is an important concern.

Author: OpenTelemetry Logs have two places to put structure (attributes and body), while often logging libraries only have one layer of structure, which makes it non-obvious (at least to me) how to do a two-way mapping between them in this case.

Also, I'd rather be overgenerous in adding open questions / future possibilities (and not worry about them until we agree on the basics).

Reviewer 15: > OpenTelemetry Logs have two places to put structure (attributes and body), while often logging libraries only have one layer of structure, which makes it non-obvious  how to do a two-way mapping between them in this case.

Maybe this should be added to the question.

> The state of this OTEP represents the option that we think will be the least confusing to the most number of users across the wide range of different language ecosystems that are supported.

Reviewer 6: 💯 agree to this phrasing! It aptly acknowledges the "naming" debate and agrees we are picking the least confusing option.

> * How to support routing logs from the Log API to a language-specific logging library

Reviewer 6: I think its best to add one more here "Should name be promoted to a top-level field in LogRecord or should it be kept as an attribute?". This was discussed several times in this OTEP too, so its best to not forget about it when we land spec.

Author: added

* How do event bodies interoperate with Span Events?

## Future possibilities

* The Event API will likely need an `IsEnabled` function based on severity level, scope name, and event name.

> * Ergonomic improvements to make it more attractive as a replacement for generic logging APIs.

Reviewer 3: "replacement" - does this mean it will have all the capability of "generic logging APIs"?

Author: I simplified this to just "Ergonomic improvements"

* Capturing raw metric events as opposed to aggregating and emitting them as OpenTelemetry Metric data

>   (e.g. [opentelemetry-specification/617](https://github.com/open-telemetry/opentelemetry-specification/issues/617)).

Reviewer 3: This sounds wrong to me. From instrumentation perspective, I think we want users to stick with an API rather than switching among APIs based on how they want to consume the data. For example, there are cases where folks want both aggregated metrics and raw measurements - in this case we do not want them to instrument twice with metrics and logging APIs.

Author: I'd prefer to be overgenerous in adding open questions / future possibilities (and not worry about them until we agree on the basic vision).

Author: (btw this came from @jsuereth and I don't think it is suggesting to instrument twice but rather to emit raw events from behind the metric API)

Reviewer 3: If it is open question, 💯 agree. I just want to avoid folks reading this as if it is the direction that we agreed.

Author: it's under the "future possibilities" section, I'm interpreting that to be a _possibility_, not something we have agreed on, so you think that needs more clarification in the text?

Reviewer 3: I might be confused by "open questions" and "future possibilities" - I thought "open questions" are something we haven't decided, "future possibilities" are something that we generally agreed on but haven't funded due to capacity reasons.

Author: got it, I've moved everything to "open questions" to avoid any ambiguity

# Event Basics

## Motivation

The introduction of Events has been contentious, so we want to document and agree on a few basics.

### What are OpenTelemetry Events?

OpenTelemetry Events are a type of OpenTelemetry Log that requires an event name and follows a specific structure implied by that event name.

They are a core concept in OpenTelemetry Semantic Conventions.

### OTLP

Since OpenTelemetry Events are a type of OpenTelemetry Log, they share the same OTLP log data structure and pipeline.

### API

> OpenTelemetry should provide a (user-facing) Logs API that includes the capability to emit OpenTelemetry Events.

Reviewer 16: My agreement to this is conditioned on the user-facing Logs API remaining simple as it is now.

Author: I added this note to the open questions around Logs API capabilities and ergonomic improvements: https://github.com/open-telemetry/oteps/pull/265/commits/3c8c0abaff75a46cf58ff1aacea9e0a0092dd1a2

Interoperability between Events and Span Events will be defined in the short term.

### SDK

This refers to the existing OpenTelemetry Log SDK.

## Alternatives

Many alternatives were considered over the past 2+ years.

These alternatives primarily boil down to differences in naming (e.g. whether to even use the word Event)
and organization (e.g. whether Event API should be something separate from Log API).

The state of this OTEP represents the option that we think will be the least confusing to the most number of users across the wide range of different language ecosystems that are supported.

## Open questions

* How to support routing logs from the Log API to a language-specific logging library
  while simultaneously routing logs from the language-specific logging library to an OpenTelemetry Logging Exporter?
* How do log bodies interoperate with other logging libraries?
  OpenTelemetry Logs have two places to put structure (attributes and body), while often logging libraries only have one layer of structure,
  which makes it non-obvious how to do a two-way mapping between them in this case.
* How do event bodies interoperate with Span Events?

> * Should the Log API have an `IsEnabled` function based on severity level, scope name, and event name?

Reviewer 15: ```suggestion
* Should the Log API have an `Enabled` function based on severity level and event name?
```

1. Logger has [`Enabled`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/bridge-api.md#enabled). It is needed also for implementing some log bridges
2. scope name is "indirectly" passed because user is providing it when creating a Logger using https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/bridge-api.md#get-a-logger

# Event Basics

## Motivation

The introduction of Events has been contentious, so we want to document and agree on a few basics.

### What are OpenTelemetry Events?

OpenTelemetry Events are a type of OpenTelemetry Log that requires an event name and follows a specific structure implied by that event name.

They are a core concept in OpenTelemetry Semantic Conventions.

### OTLP

Since OpenTelemetry Events are a type of OpenTelemetry Log, they share the same OTLP log data structure and pipeline.

### API

> OpenTelemetry should provide a (user-facing) Logs API that includes the capability to emit OpenTelemetry Events.

Reviewer 5: Is "should" normative here? For languages implementations that it makes sense to wrap already existing Log APIs can they ignore this recommendation?

Reviewer 15: It looks like it should be "must" as otherwise it would be against what is written a little lower in "Alternatives" section

> These alternatives primarily boil down to differences in [...]
organization (e.g. whether Event API should be something separate from Logs API).

Reviewer 17: IIRC we don't use normative language _as much_ in the OTEPs repo (just FYI).

Reviewer 5: This needs to be made clear what the expectation is.

We can't just assume readers are going to understand this as requirement when we are communicating it as a pseudo-recommendation.

I do not think it should be a requirement. And if that is what is trying to be merged here it needs to be made clear.

Reviewer 15: Given the amount of discussions around Events API vs Logs API, I agree that this section should be clarified. I feel that many reviewers (including myself) accepted the OTEP, because of the conversation and "implicit" agreements made during SIG meetings. These conclusions could be better described.

Author: as discussed in Spec SIG meeting, I believe this is resolved now by changing should to SHOULD in https://github.com/open-telemetry/oteps/pull/265/commits/e46600340cdd0344dde4387f63219b2c7459f5a7

resolving this discussion, but please open a new one with summary if there are any additional concerns beyond that, thanks!

# Event basics

## Motivation

The introduction of Events has been contentious, so we want to write down and agree on a few basics.

### What are events?

Events are a type of log which has a required event name and a specific structure which is implied by that event name.

The event name and its implied structure make events much more suitable for observability compared to generic logs.

Because of this, the OpenTelemetry project wants to encourage the transition from generic logs to events.

### OTLP

Since events are a type of log, they share the same OTLP data structure and OTLP pipeline.

### API

> OpenTelemetry should have an Event API. This will help to promote the distinction between events and generic logs,

Reviewer 11: ```suggestion
The OpenTelemetry logger should include methods for emitting events as well as application logs. The term "event" helps to promote the distinction between these two use cases. Methods such as `EmitEvent` should be used for instrumentation, while methods such as `EmitLog` should be used for consuming application logs.
```

Reviewer 14: I agree with Nev that it's very confusing as a customer to use my logger to emit Real User Monitoring (RUM) events. It's a much cleaner story to keep these separate from simple info logs. They are conceptually two very different things. It feels like we're shoehorning in a completely separate concept into an existing construct for some short-term gains. I suggest we stop and ask customers whether they consider a click-event as an enhanced diagnostic log or as a separate concept altogether.

Reviewer 11: Application developers do not really care about these details. I have talked to many developers over the past two years, and they are not bothered by phrases such as "log an event" or "events are stored as logs." If you try to get application developers to dig deeply into this subject they quickly lose interest, it is very clear that it does not stir a lot of passions and they do not really care about the details. 

The only thing that consistently bothered users was the proposal that we not provide an API. When asked to make an event, pretty much all users wanted a line of code to copy and paste. Whether that line of code says `eventLogger.EmitEvent()` or `logger.EmitEvent()` doesn't really matter to them.

When it comes to RUM customers, there is no need for a RUM system to refer to logs. A RUM system also doesn't need to refer to spans, or OpenTelemetry at all. The RUM domain is specific to the browser, so the language used in the RUM UI should match the language used by the browser community to describe browser operations.

The same goes for GenAI. The internal details of OpenTelemetry are not what a specialized GenAI analysis tool designs its UI around. They design using the language of GenAI, aka the semantic conventions. It doesn't really matter what the transport protocol happens to name the data structures, or which API that developers chose to use to generate the data.

Reviewer 6: > Whether that line of code says eventLogger.EmitEvent() or logger.EmitEvent() doesn't really matter to them.

+1. Most people would prefer a helper (possibly auto-generated using weaver/similar) method like `EmitButtonClick(buttonname,username,etc.)`

Reviewer 11: 100% agree that the next step in API design for Otel should be to create more of these high level "instrumentor" helpers.

Author: Resolving this based on latest updates. If you'd like to continue discussing, please start a new thread with a summary of any remaining points you'd like to discuss. Thanks.

Note: If a user chooses to send events from the Event API to a generic logging library, and they have
also chosen to send the logs from their generic logging library to the OpenTelemetry Logging SDK, then they should
avoid sending events from the Event API directly to the OpenTelemetry Logging SDK since that would lead to duplicate
capture of events that were sent from the Event API.

It should be possible to bypass the Event API entirely and emit Events directly via an existing language-specific logging libraries.
This helps reinforce the idea that events are just a specific type of log.

Note: Because of this, generic event processors should be implemented as Log SDK processors.

OpenTelemetry will recommend that
[instrumentation libraries](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/glossary.md#instrumentation-library)
use the Event API over emitting events via a generic logging library in order to give users a simple and consistent
onboarding story that doesn't involve mixing the two approaches.

OpenTelemetry will recommend that application developers also use the Event API over emitting events via a generic

> logging library since it avoid accidentally emitting logs which lack an event name or are unstructured.

Reviewer 18: ```suggestion
logging library since it avoids accidentally emitting logs which lack an event name or are unstructured.
```

Also, recommending the Event API over emitting events via a generic logging library makes for a clearer overall
OpenTelemetry API story - with first class user facing APIs for traces, metrics, and events,
all suitable for using directly in native instrumentation.

### Relationship to Span Events

Events are intended to replace Span Events in the long-term.
Span Events will be deprecated to signal that users should prefer Events.

Interoperability between Events and Span Events will be defined in the short-term.

### SDK

The Event SDK needs to support two destinations for events:

* OpenTelemetry Logging SDK
* Language-specific logging libraries

## Trade-offs and mitigations

TODO

## Prior art and alternatives

> TODO

Reviewer 15: Can please describe why instead of having "Event API" we cannot simply use "Logs (Bridge) API" directly? The logs/events would simple have attributes (e.g. `event.name`) and instrumentation scope as defined in semantic conventions.

Reviewer 7: > Can you please describe why instead of having "Event API" we cannot simply use "Logs (Bridge) API" directly?

That's an interesting idea. The components in the log bridge API (LoggerProvider, Logger, etc) were at least partially named in a way that would make sense in a future world where the API had a scope expanded to be user facing (i.e. no longer just a bridge).

Today `Logger` has a single operation: [emit a LogRecord](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/bridge-api.md#emit-a-logrecord)

You could imagine adding an additional operation for `emit an Event` and adjusting the phrasing so that `emit a LogRecord` is non-user user facing, reserved for bridging, and `emit an Event` is user facing.

Reviewer 15: I could also imagine a method on `LogRecord` like `asEvent(string name)` that would be just adding an `event.name` string attribute.

However, for sake of instrumentation libraries, I do not think we even need that. What is a problem just adding the attribute explicitly like all other attributes?

I was thinking that it could nice for "business code" so that they can easily emit events. However, even then just making a "facade" for adding event name would not give much if the user is not familiar with the concepts of semantic conventions and schemes. For people understanding these concepts adding an `event.name` attribute  should not be an issue. 

Also, I do not think we should "restrict" semantic logging to only events.

Reviewer 4: I think the moment we document `Logger` as user-facing, we can no longer keep one of its methods non-user facing (but technically public). 

With this proposal, we'd make all logging API public and with @jsuereth analogy https://github.com/open-telemetry/oteps/pull/265#discussion_r1742560611 it's not a bad thing at all.

Reviewer 15: nit: `Logger` is already user-facing. However, it is should be used in log bridges. Excerpt Go docs:

> Package log provides the OpenTelemetry Logs Bridge API.
>
> This package is intended to be a bridge between existing logging libraries and OpenTelemetry. It is not designed to be a logging API itself.

But nothing prevents developers from using it.

Reviewer 15: I also fear the confusion that we would introduce for users when Logs SDK would be used to e.g. process events emitted via Events API.

Reviewer 4: > I also fear the confusion that we would introduce for users when Logs SDK would be used to e.g. process events emitted via Events API.

could you elaborate? events are logrecords with an extra attribute so there should be no surprises.

Reviewer 15: > events are logrecords with an extra attribute so there should be no surprises.

Do you think it would be obvious for end-users who just want to emit events using Event API?

If this would be "no surprise" then it would be totally fine to ask end-users to add the attribute by themselves.

Reviewer 4: So the concern is that if we have Event API, then it's logical to have Event SDK so it's less surprising that events are logs with extra attribute?

I assume having `Logger.emitLog` and `Logger.emitEvent` would resolve it?

Reviewer 15: > I assume having Logger.emitLog and Logger.emitEvent would resolve it?

Yes, if it is in Logs API (and not Events API).

However, I do not find it worth to add a method where the only difference from `emitLog` would be that it also adds an attribute.

Reviewer 15: Moreover, what would we do with events produced via logging libraries which are bridged via Logs API (logs with explicitly added `event.name` attribute/field) ? Aren't they also events?

Reviewer 4: >  libraries which are bridged via Logs API (logs with explicitly added event.name attribute/field) ? Aren't they also events

yes, they are events according to event definition

> If a logging library is capable of creating logs which correctly map
> to the Event data model, logging in this manner is also an acceptable way to
> create Events.

https://github.com/open-telemetry/opentelemetry-specification/blob/625e0f8500f2aea48982a943997c991ff41ee9bb/specification/logs/event-api.md?plain=1#L79

Reviewer 19: If we only need to publish `LogRecord` with different flavors of parameters, like publishing events, having a dedicated provider or API doesn't make much sense to me.

This is the current design:

```mermaid
graph TD
    A[LoggerProvider] -->|Get| B(Logger)
    B -->|Emit| C(LogRecord)

    D[EventLoggerProvider] -->|Get| E(EventLogger)
    E -->|Emit| F(LogRecord)
```

I propose we get rid of `EventLoggerProvider`:

```mermaid
graph TD
    A[LoggerProvider] -->|Get| B(Logger)
    B -->|Emit| C(LogRecord)

    A -->|Get| E(EventLogger)
    E -->|Emit| F(LogRecord)
```

Or, even better, take the `EventLogger` as a wrapper around the `Logger` (I prefer this):

```mermaid
graph TD
    A[LoggerProvider] -->|Get| B(Logger)
    B -->|Emit| C(LogRecord)

    B -. Wrapped .-> E(EventLogger)
    E -->|EmitEvent| C
```

The pseudocode would be:

```go
lp = log.NewLoggerProvider()
otel.SetLoggerProvider(lp)

l = lp.Logger()

el := EventLogger(l)
el.EmitEvent()
```

This looks simple to understand and easy to maintain. Maintainers can focus more on existing log implementation instead of setting up a new set of APIs.

Reviewer 3: 💯 👍

Reviewer 4: leaving naming aside, why not

```mermaid
graph TD
    A[LoggerProvider] -->|Get| B(Logger)
    B -->|Emit| C(LogRecord)
    B -->|EmitEvent| C
```

?

Reviewer 19: > leaving naming aside, why not

This is also good, as it might improve performance by reducing the calling chain. However, the `Logger` is a bridge that is not intended to be called by application developers directly. Having `Emit` aside from `EmitEvent`, intended to be called by application developers, feels strange to me.

> The Logs Bridge API is not intended to be called by application developers directly, and SHOULD include documentation that discourages direct use.

https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#artifact-naming

This approach also needs to change [Logs Bridge API](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/), a stable doc, to include `EmitEvent`.

---

The wrapper design can separate `Emit` and `EmitEvent`. So, we do not need to touch the `Logs Bridge API`; leave the `Event API` to only have `EventLogger` to state, and it is a wrapper of `Logger`.

---

Both designs are good to me. I slightly prefer the wrapper one.

Reviewer 15: > Or, even better, take the `EventLogger` as a wrapper around the `Logger` (I prefer this):

Indeed a lot better than adding Event API 👍 However, why not simply?

```go

lp = log.NewLoggerProvider()
otel.SetLoggerProvider(lp)

l = lp.Logger()

el := EventLogger(l)
ev := Event{} // or ev := Record{}? would we need to add Event type which would be almost the same as Record?
el.EmitEvent("special-event", ev)
```

vs

```go
lp = log.NewLoggerProvider()
otel.SetLoggerProvider(lp)

l = lp.Logger()

ev := Record{}
ev.AddAttribute(semconv.EventName("special-event")) r.AsEvent
l.Emit(ev)
```

Alternatively, we can add a method/helper function which makes a log record an event e.g.

```go
lp = log.NewLoggerProvider()
otel.SetLoggerProvider(lp)

l = lp.Logger()

ev := Record{}
ev.AsEvent("special-event")
l.Emit(ev)
```

Reviewer 15: > leaving naming aside, why not

Instrumentation libraries want emit events and not  "raw logs". Thus they will use EventLogger. They do not need `Emit`.
Log bridges want to emit "raw logs" - they do not need `EmitEvent`.
Also, the API implementation would not need to implement `EventLogger` (as this is just a facade). 

Loose reference: https://martinfowler.com/bliki/RoleInterface.html

Reviewer 19: > However, why not simply

Yeah, it seems more straightforward. After all, developers can access the `Logger` if they want.

Then, we can add a helper function to construct an "Event" later based on user feedback.

---

I feel that by **removing this statement**, we can solve all of these Log issues. 🤣

> The Logs Bridge API is not intended to be called by application developers directly, and SHOULD include documentation that discourages direct use.

https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#artifact-naming

Reviewer 7: > I feel that by removing this statement, we can solve all of these Log issues. 🤣

The log bridge API elements were named to allow for evolution into a user facing log API if needed. So this definitely seems feasible and an allowable change. From that [link](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/bridge-api.md#artifact-naming):

> in the event OpenTelemetry were to add a user facing API, the Logs Bridge API would be a natural starting point. Therefore, Log Bridge API artifact, package, and class names MUST NOT include the terms "bridge", "appender", or any other qualifier that would prevent evolution into a user facing API.

Reviewer 11: The Event API and the LogBridge API should stay separate, as they represent different parts of our logging system.

Otel has two goals when it comes to logging:

* Consume all of the end users' existing logs.
* Produce our own highly opinionated logs, which will replace span events.

These goals are at the opposite ends of the spectrum. When consuming existing logs we want to be as flexible as possible, when creating our own logs we want to be as rigorous as possible. That's why we have both the LogBridge API and the Event API.

The LogBridge is intended to be used as a log sink for exporting logs of any kind. This includes the events – by default, the Event SDK uses the LogBridge as their event exporter. Optionally, users may want to use SpanEvents as their event exporter. Or they may not want to do any processing at all, and just send the events directly to their application logging system.

Keeping the EventProvider and LoggerProvider separate from each other avoids most of the common pitfalls relating to circular log pipelines and dependencies. For example, users may want to configure their application logger to export data to Otel via the LogBridge. At the same time, they may want all Otel instrumentation (which uses events) to be configured to send data directly to their application logger.  So events flow into the application logger, then all logs flow into the log bridge. Alternatively, events might need a different logging path from regular application logs, in order to preserve their structure and avoid unnecessary processing. Having separate configurations and processor pipelines for the Event API and LogBridge API makes it possible for end users to stay in control of their processing.

Reviewer 15: > when creating our own logs we want to be as rigorous as possible

is it not the purpose of semantic conversations?

> Having separate configurations and processor pipelines for the Event API and LogBridge API makes it possible for end users to stay in control of their processing.

Then you suggest that these should have separate SDKs? Wouldn't this contradict with the idea that events may be emitted by using logging libraries as noted in https://github.com/open-telemetry/oteps/pull/265#discussion_r1745891158. Similar things could be achieved by creating a routing log record processor without having the need to create new API and SDK.

Reviewer 10: Promoting the existing logging APIs causes a few problems, long-term, I think.

First, and most problematic imo, is that it dramatically increases the 'surface area' of the SDK from a maintainability perspective without actually giving us much in return. The point of the bridge is that if you have legacy logging APIs, you can get more out of those logs by feeding them into the SDK to be associated with context or get normalized attributes. If we walk that back and say 'well, ok, do whatever' then we'd need to add more features to be competitive with existing logging APIs and SDKs.

The second is that we would be promoting logs as a signal that we think people should use. This is a bigger mistake, imo. Traditional logging patterns are, frankly, not terribly useful in the OpenTelemetry model. Events are. API separation makes that crystal clear.

Reviewer 11: trying to summarize here and bring these threads together into a decision: https://github.com/open-telemetry/oteps/pull/265#issuecomment-2334806152

Reviewer 15: Shouldn't  we put a brief description about the alternative of having a separate API and SDK any why it is rejected?

Author: done: https://github.com/open-telemetry/oteps/pull/265/commits/fc0381e8d9710cb7afccff2b7eb81fe921aacee3
